server/src/main/java/org/elasticsearch/transport/netty4/HostBasedAuthHandler.java

```diff
@@ -66,22 +63,22 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
         String userName = Role.CRATE_USER.name();
         var authMethod = authentication.resolveAuthenticationType(userName, connectionProperties);
         if (authMethod == null) {
-            ReferenceCountUtil.release(msg);
-            authError = new AuthenticationException("No valid auth.host_based entry found for: " + remoteAddress);
-            CloseableChannel tcpChannel = ctx.channel().attr(Netty4Transport.CHANNEL_KEY).get();
-            CloseableChannel.closeChannel(tcpChannel, true);
-            throw authError;
+            closeAndThrowException(ctx, msg, new AuthenticationException("No valid auth.host_based entry found for: " + remoteAddress));
         }
         try {
             authMethod.authenticate(userName, null, connectionProperties);
             ctx.pipeline().remove(this);
             super.channelRead(ctx, msg);
         } catch (Exception e) {
-            ReferenceCountUtil.release(msg);
-            authError = e;
-            CloseableChannel tcpChannel = ctx.channel().attr(Netty4Transport.CHANNEL_KEY).get();
-            CloseableChannel.closeChannel(tcpChannel, true);
-            throw e;
+            closeAndThrowException(ctx, msg, e);
         }
     }
+
+    private void closeAndThrowException(ChannelHandlerContext ctx, Object msg, Exception e) throws Exception {
+        ReferenceCountUtil.release(msg);
+        authError = e;
```

Comment:
I'd keep this assignment outside to avoid the side effect. It's not clear from the naming that it does that. Then it should also be possible to make the method static

Reply:
Cannot make it static if I throw inside since authError is not static. 
Or you meant assign and throw outside?

Reply:
Moved outside but didn't make the method static - otherwise would have to make `authError` static as well which is dangerous as we can create multiple instances of `HostBasedAuthHandler`.

Reply:
It should throw `e` instead

Reply:
updated